### PR TITLE
Allow reusing and reseting wires after a mid-circuit measurement

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
Making mid-circuit measurements more feature rich.

**Description of the Change:**
* `qml.defer_measurements` no longer raises an error if a qubit is reused after measurement, and includes functionality to reset qubits.
* Updated `qml.measure` to include a `reset` keyword argument for reseting qubits to the $|0\rangle$ computational basis state after measurement.

Qubit reuse is enabled by by applying a `CNOT` gate controlled on the measured wire to an auxiliary target wire to store the measurement state on the auxiliary wire. Using `CNOT` will cause entanglement between the measured wire and the rest of the wires to be "copied" to the auxiliary wire. Hence, the measured wire can be reused while still preserving the measurement state in the auxiliary wire, which, `defer_measurements` uses as the control wires for controlled operations created by `qml.cond`.

For qubit reset, `qml.defer_measurements` applies a `CNOT` gate controlled on the auxiliary wire to the target measured wire. A `CNOT` between two wires in the same state reduces the target wire to the $|0\rangle$ state.

**Benefits:**
* Qubit reuse and reset is possible on PennyLane

**Possible Drawbacks:**
* Until `defer_measurements` is ported over to the new transforms API, custom wire labels will not be allowed.
* Users need to remember to initialize old-gen devices with n extra wires for n mid-circuit measurements, and make sure that the **_last_** n wires are reserved for mid-circuit measurement states.
* Space complexity for circuits with mid-circuit measurements will scale as $O(2^n)$ for n mid-circuit measurements, as a new wire will be added for each mid-circuit measurement.
* Since new wires will be added, old-gen devices will need to be initialized with as many extra wires as the number of mid-circuit measurements.

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mudit2812** [comment]
[sc-36565] [sc-38882]

File: pennylane/transforms/defer_measurements.py [hunk 1]
```diff
@@ -84,34 +92,65 @@ def qfunc(par):
...
+    new_wires = {}
 
+    for op in tape.circuit:
         if isinstance(op, MidMeasureMP):
-            measured_wires[op.id] = op.wires[0]
+            new_wires[op.id] = cur_wire
+            qml.CNOT([op.wires[0], cur_wire])
+
+            if op.reset:
+                qml.CNOT([cur_wire, op.wires[0]])
```
Comment (trbromley) [inline pennylane/transforms/defer_measurements.py:136]: Shouldn't `op.reset is True` indicate that we don't need to apply any CNOTs at all?

Comment (mudit2812) [reply pennylane/transforms/defer_measurements.py:136]: I don't follow. Measurement causes measured state (let's say on wire 0) to be stored in a new wire (say 3) using `CNOT([0, 3])`. If `op.reset is True`, then we want wire 0 back in $|0\rangle$, which we would achieve by applying `CNOT([3, 0])`.

File: pennylane/measurements/mid_measure.py [hunk 2]
```diff
@@ -24,8 +24,8 @@
 from .measurements import MeasurementProcess, MidMeasure
 
 
-def measure(wires):  # TODO: Change name to mid_measure
-    """Perform a mid-circuit measurement in the computational basis on the
+def measure(wires: Wires, reset: Optional[bool] = False):  # TODO: Change name to mid_measure
```
Comment (albi3ro) [inline pennylane/measurements/mid_measure.py:27]: Is this TODO ever going to happen?

Comment (mudit2812) [reply pennylane/measurements/mid_measure.py:27]: Lol. I can add it to this PR if needed

File: pennylane/measurements/mid_measure.py [hunk 3]
```diff
@@ -55,23 +55,72 @@ def func(x, y):
...
+            m_0 = qml.measure(1, reset=True)
+
+            qml.cond(m_0, qml.RY)(y, wires=0)
+            qml.RX(np.pi/4, wires=1)
+            return qml.probs(wires=[0, 1])
+
+    Executing this QNode:
+
+    >>> pars = np.array([0.643, 0.246], requires_grad=True)
+    >>> func(*pars)
```
Comment (albi3ro) [inline pennylane/measurements/mid_measure.py:78]: Might be worth finding an example where we can clearly see the effect of `reset` on the outcome.

Maybe something along the lines of:
```
dev = qml.device("default.qubit", wires=3)
@qml.qnode(dev)
def func():
    qml.PauliX(1)
    m_0 = qml.measure(1, reset=reset)
    return qml.probs(wires=[1])
```

File: pennylane/transforms/defer_measurements.py [hunk 4]
```diff
@@ -84,34 +92,60 @@ def qfunc(par):
...
-        op_wires_measured = set(wire for wire in op.wires if wire in measured_wires.values())
-        if len(op_wires_measured) > 0:
-            raise ValueError(
-                f"Cannot apply operations on {op.wires} as the following wires have been measured already: {op_wires_measured}."
-            )
+    # Current wire in which pre-measurement state will be saved if dev_wires not specified
+    cur_wire = max(tape.wires) + 1
+    new_wires = {}
 
+    for op in tape.circuit:
```
Comment (albi3ro) [inline pennylane/transforms/defer_measurements.py:130]:
```suggestion
    for op in tape:
```

File: pennylane/transforms/defer_measurements.py [hunk 5]
```diff
@@ -84,34 +92,60 @@ def qfunc(par):
...
+            if op.reset:
+                qml.CNOT([cur_wire, op.wires[0]])
+
+            cur_wire += 1
 
         elif op.__class__.__name__ == "Conditional":
-            _add_control_gate(op, measured_wires)
+            _add_control_gate(op, new_wires)
         else:
             apply(op)
```
Comment (albi3ro) [inline pennylane/transforms/defer_measurements.py:143]:
```suggestion
            apply(op)
            return tape._qfunc_output
```

File: pennylane/transforms/defer_measurements.py
Comment (albi3ro) [inline pennylane/transforms/defer_measurements.py:1]: If nothing else uses the wire again, do we need to use a bonus auxiliary wire? Or could we only add an auxiliary wire if something needs that wire after the measurement?

Comment (mudit2812) [reply pennylane/transforms/defer_measurements.py:1]: That would only be the case if someone makes a measurement but doesn't reset the measured wire or uses the measurement value for `qml.cond`. I don't really see users doing that often.

Comment (mudit2812) [reply pennylane/transforms/defer_measurements.py:1]: Moreover, the intention with changing `defer_measurements` is to have one transform that works for every device. Further optimizations can be added for device-specific implementations. For example, much of the functionality offered by mid-circuit measurements can be achieved by somehow making use of the probabilities of the measurement outcomes (eg, postselection, measurement statistics), so a device specific implementation could be to use the state vector to compute and store the marginal probabilities in a local register when classical control or qubit resetting is not occurring. These would come later though.

File: pennylane/measurements/mid_measure.py (hunk 2 above)
Comment (trbromley) [reply pennylane/measurements/mid_measure.py:27]: I don't think we need to change to `mid_measure` :+1:

File: pennylane/transforms/defer_measurements.py
Comment (mudit2812) [reply pennylane/transforms/defer_measurements.py:1]: After some offline conversation with @timmysilv I've realized that we could definitely avoid adding a new wire if the measured wire is not being reused. I don't want to delay merging this PR any further, so I'd like to add this optimization in a future PR.

File: pennylane/measurements/mid_measure.py (hunk 2 above)
Comment (mudit2812) [reply pennylane/measurements/mid_measure.py:27]: In that case I'll remove the `TODO`

File: pennylane/transforms/defer_measurements.py (hunk 1 above)
Comment (mudit2812) [reply pennylane/transforms/defer_measurements.py]: I've opened another PR that will address this change after 0.32 is released. For now, the current implementation will remain.

File: tests/transforms/test_defer_measurements.py [hunk 6]
```diff
@@ -902,13 +985,120 @@ def qnode2(parameters):
...
+        @qml.qnode(dev)
+        def qnode(p, x, y):
+            qml.Hadamard(0)
+            qml.PhaseShift(p, 0)
+            # Set measurement_ids so that the order of wires in combined
+            # measurement values is consistent
+
+            mp0 = qml.measurements.MidMeasureMP(0, reset=True, id=0)
+            m0 = qml.measurements.MeasurementValue([mp0], lambda v: v)
+            qml.cond(~m0, qml.RX)(x, 1)
```
Comment (lillian542) [inline tests/transforms/test_defer_measurements.py:1058]: What does ~ do?

**Review by lillian542 — APPROVED** [review]
Looks good! 🚀 I left a few very minor suggestions/questions wrt tests, but that's all from me.

File: tests/transforms/test_defer_measurements.py (hunk 6 above)
Comment (mudit2812) [reply tests/transforms/test_defer_measurements.py:1058]: It's a binary NOT operator

File: tests/test_qnode.py [hunk 7]
```diff
@@ -1077,7 +1077,7 @@ def test_defer_meas_if_mcm_unsupported(self, first_par, sec_par, return_type):
         """Tests that the transform using the deferred measurement principle is
         applied if the device doesn't support mid-circuit measurements
         natively."""
-        dev = qml.device("default.qubit", wires=2)
+        dev = qml.device("default.qubit", wires=3)
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply tests/test_qnode.py:1080]: I added the extra wire before I optimized the transform to not add new wires if the wire was not reused. Forgot to remove it, but it makes no difference to the outcome anyway.

File: tests/devices/qubit/test_preprocess.py [hunk 8]
```diff
@@ -217,7 +217,7 @@ def test_expand_fn_expand_unsupported_op(self, shots):
     # pylint: disable=no-member
     def test_expand_fn_defer_measurement(self):
         """Test that expand_fn defers mid-circuit measurements."""
-        mp = MidMeasureMP(wires=[0], id="test_id")
+        mp = MidMeasureMP(wires=[0], reset=True, id="test_id")
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply tests/devices/qubit/test_preprocess.py:220]: Whether the circuit is expanded correctly with `defer_measurements` is tested in the `test_defer_measurements.py` file so this test is just to verify that the transform is used, not to check its correctness.

File: pennylane/transforms/defer_measurements.py [hunk 9]
```diff
@@ -40,6 +43,14 @@ def defer_measurements(tape):
         that can be controlled as such depends on the set of operations
         supported by the chosen device.
 
+    .. note::
+
+        Devices that inherit `QubitDevice` **must** be initialized with an additional
```
Comment (timmysilv) [inline pennylane/transforms/defer_measurements.py:48]:
```suggestion
        Devices that inherit from :class:`~pennylane.QubitDevice` **must** be initialized with an additional
```

File: pennylane/transforms/defer_measurements.py [hunk 10]
```diff
@@ -40,6 +43,14 @@ def defer_measurements(tape):
...
+        wire for each mid-circuit measurement after which the measured wire is reused
+        or reset for `defer_measurements` to transform the quantum tape correctly.
```
Comment (timmysilv) [inline pennylane/transforms/defer_measurements.py:50]:
```suggestion
        or reset for ``defer_measurements`` to transform the quantum tape correctly.
```

File: pennylane/transforms/defer_measurements.py [hunk 11]
```diff
@@ -40,6 +43,14 @@ def defer_measurements(tape):
...
+        Such devices should also be initialized without custom wire labels for correct
+        behaviour.
```
Comment (timmysilv) [inline pennylane/transforms/defer_measurements.py:52]: i see that you use `max(wires) + 1`. can this comment be generalized for new devices that don't need wires either? they'd still need a _circuit_ with standard wire labels

Comment (mudit2812) [reply pennylane/transforms/defer_measurements.py:52]: Yep, totally can. I still have grievances about this behaviour though. I was thinking of just using `qml.map_wires` but I don't like it because if we have a tape with `wires=[1, 0, 3, 2]` (which is totally possible depending on which wires are used first in the queue), they would get mapped to `wires=[0, 1, 2, 3]`, which would be the same labels but permuted, and the user might not realize that the labels have changed.

**Review by timmysilv — APPROVED** [review]

File: pennylane/transforms/condition.py [hunk 12]
```diff
@@ -114,6 +114,8 @@ def qnode(x, y):
         Expressions with boolean logic flow using operators like ``and``,
         ``or`` and ``not`` are not supported as the ``condition`` argument.
 
+        While such statements may not result in errors, they may result in
+        incorrect behaviour.
```
Comment (trbromley) [inline pennylane/transforms/condition.py:118]: What does this mean?

**Review by trbromley — COMMENTED** [review]
Thanks @mudit2812! Will we have a follow up PR to update the docs [here](https://docs.pennylane.ai/en/stable/introduction/measurements.html#mid-circuit-measurements-and-conditional-operations)?
